### ISSUE TITLE
compiling resources

### ISSUE THREAD

**Issue by hdd** [mottosso/Qt.py#88]
Hi , is there any wrapped command for things like : **pyside-rcc** and **pyside-uic** , so does pick up the right one depending on the framework in use ?

**Comment by hdd** [comment]
wondering, if the tool has to work on more than one (eg: PySide/PySide2) ... should resource should be compiled for both ?

**Comment by hdd** [comment]
checking now, but should be compatible, so is matter of having some logic on when we compile the resource files.

**Comment by hdd** [comment]
ok looks like is not, as in the compiled resouce file  we end up (eg: pyside) with from 

PySide import QtCore

**Comment by mottosso** [comment]
There isn't any wrapper at the moment, but one could potentially be built in.

I reckon that if the resulting resources can't be used in all bindings, then there probably isn't much point in wrapping it.

If you were to provide a wrapper, it could get tested on each supported platform and each supported binding via Travis and the pull request.  Have a look at the currently ongoing pull request for load_ui for example.

**Comment by fredrikaverpil** [comment]
To my knowledge, `pysideuic` is its own module and is not part of `PySide` or `PySide2`. So far, we haven't attempted to map modules which aren't part of the actual bindings but I can't see why this could not be done.

How would such a module be imported though?
Perhaps something like this:

    from Qt import pysideuic

...?

**Comment by fredrikaverpil** [comment]
Oh, wait, are we talking about the standalone executables such as `pyside-uic.exe`?

**Comment by hdd** [comment]
Thinking of fixing this in the various setup.py which are doing the
compilation  of the resources, replacing in the compiled resource :

<watever>.QtGui with Qt.QtGui

**Comment by mottosso** [comment]
> Thinking of fixing this in the various setup.py which are doing the
compilation  of the resources, replacing in the compiled resource

Possible to show an example of this?

**Comment by hdd** [comment]
Sure, so when you compile an py from an rcc file, the first lines of the compiled one looks like this:

```
# -*- coding: utf-8 -*-

# Resource object code
#
# Created: Fri Aug 5 10:17:32 2016
#      by: The Resource Compiler for PySide (Qt v4.8.5)
#
# WARNING! All changes made in this file will be lost!

from PySide import QtCore
qt_resource_data = "......................."
```

This should be changed after the compilation as : 

```
# -*- coding: utf-8 -*-

# Resource object code
#
# Created: Fri Aug 5 10:17:32 2016
#      by: The Resource Compiler for PySide (Qt v4.8.5)
#
# WARNING! All changes made in this file will be lost!

from Qt import QtCore
qt_resource_data = "......................."
```

I'll give this  a shot and let you know.
This way we should be able to provide backward compatibility.

**Comment by hdd** [comment]
so, in the setup.py of the tool I'm working in at the moment, I've been adding a post processing call which does pretty much this, which seems to be working fine :

```
import fileinput
filepath = '/home/efesto/Desktop/resource.py'
replace = 'from Qt import QtCore'
for line in fileinput.input(filepath, inplace=True):
	if 'import QtCore' in line:
		print line.replace(line, replace)
	else:
		print line
```

**Comment by hdd** [comment]
wondering, does worth providing a wrapper with popen which kicks the compilation and replace ?
it could be retained in a /bin folder under the project as qt-rcc

let me know.
Cheers.
L

**Comment by mottosso** [comment]
Interesting. I haven't used these features myself, so I'll rely on your experiences.

I'm a bit blurry on how the setup.py comes in to play. Are you saying that you distribute your software with a PyQt or PySide compiled resource file, and re-map it when setup.py is used? That is, during installation?

Could you show a small runnable example?

**Comment by mottosso** [comment]
To cut to the chase, what I'm looking for is the tests for this feature, before we start talking about an implementation.

The tests will more or less be the minimal, yet complete, examples that you come up with. Once there are tests, it will be more easily understood what your end game is and how we can reach it.

For example.

```python
def test_compile_resources():
  import Qt
  Qt.compile_resources(dirname="/path/to/resources", fname="my_resources.py")
  assert os.path.isfile("path/to/resources/my_resources.py"), "nothing written"
```

This tells us, in just a few lines:

1. How to use the end result
2. What the interface will look like
3. What the expected result is

The tests could also include importing this file and asserting some of its content, like (1) the variable in which the data resides and (2) the header comment and (3) the use of Qt instead of PySide etc.

**Comment by mottosso** [comment]
How about this for support coming from Qt.py?

1. Compile natively with either PyQt or PySide
2. Search-and-replace `PyQt` or `PySide` with `Qt` and write a new file
3. Done

That way, developers can work natively and Qt.py doesn't have to wrap any functionality, whilst still being able to provide support for all bindings. Win-win-win.

The interface could look something like this.

```bash
$ pyside-uic my_resources # writes my_resources.py
$ python -m Qt --convert my_resources.py # writes my_resources_qt.py
```

**Comment by fredrikaverpil** [comment]
I wonder if they produce identical results though?
If so, I agree this is a win-win-win 😉

**Comment by hdd** [comment]
isn't it a bit inconsistent as you are providing wrappers for loading the ui ?
This two step process doesn't make much sense, either is done from within the code (Qt.compile_resource(in, out) ), or for me , simply worth keeping it on the various setup.py

**Comment by mottosso** [comment]
> simply worth keeping it on the various setup.py

Fair enough. Let's make it part of the docs on how to do that, and we're done.

**Comment by hdd** [comment]
cool, then please remember to add this also for the uic files, otherwise people will be having all sorts of issues.

Cheers.
L.

**Comment by mottosso** [comment]
> then please remember to add this also for the uic files

What uic files? Sorry, I don't follow.

**Comment by hdd** [comment]
Ui (uic) files are created through designer-qt4 and the can be compiled
(similarly to resources) to py using pyside-uic

This, usually, allow to load the ui faster than through the load ui
command.

**Comment by mottosso** [comment]
I see, ok. Would you be up for writing this part of the README?

**Comment by fredrikaverpil** [comment]
@hdd please have a look at the latest release which implements the alpha feature --convert: https://github.com/mottosso/Qt.py/releases/tag/0.5.0

**Comment by mottosso** [comment]
I think Lorenzo is asking about compiling resources, such as png's and jpeg's, into a .py file. `--convert` only handles .ui files.

**Comment by fredrikaverpil** [comment]
Oh 😮 allright 😄
